codegen/chain.go

```diff
@@ -232,12 +232,26 @@ func (cg *CodeGen) EmitFilesystemBuiltinChainStmt(ctx context.Context, scope *pa
 			return fc, err
 		}
 
+		fi, err := os.Stat(path)
+		if err != nil {
+			return fc, err
+		}
+
 		var opts []llb.LocalOption
 		for _, iopt := range iopts {
 			opt := iopt.(llb.LocalOption)
 			opts = append(opts, opt)
 		}
 
+		if !fi.IsDir() {
+			filename := filepath.Base(path)
+			path = filepath.Dir(path)
+
+			// When path is a filename instead of a directory, include and exclude
+			// patterns should be ignored.
+			opts = append(opts, llb.IncludePatterns([]string{filename}), llb.ExcludePatterns([]string{}))
```

Comment:
so it is the last Include/ExcludePatterns option in the list that wins?  They are not merged?

Reply:
Yeah, under the hood BuildKit just sets the array rather than appending it.

https://github.com/moby/buildkit/blob/1e4c22119456d26e7818473290962bc4503c8256/client/llb/source.go#L312-L319